util/types/json/json.go

```diff
@@ -15,26 +15,126 @@ package json
 
 import (
 	"bytes"
+	"encoding/json"
+	"fmt"
+	"strings"
+	"unsafe"
 
 	"github.com/pingcap/tidb/mysql"
 	"github.com/pingcap/tidb/terror"
+	"github.com/pingcap/tidb/util/hack"
 )
 
+const (
+	typeCodeObject  byte = 0x01
+	typeCodeArray   byte = 0x03
+	typeCodeLiteral byte = 0x04
+	typeCodeInt64   byte = 0x09
+	typeCodeFloat64 byte = 0x0b
+	typeCodeString  byte = 0x0c
+)
+
+const (
+	jsonLiteralNil   byte = 0x00
+	jsonLiteralTrue  byte = 0x01
+	jsonLiteralFalse byte = 0x02
+)
+
+const unknownTypeCodeErrorMsg = "unknown type code: %d"
+const unknownTypeErrorMsg = "unknown type: %s"
+
 // JSON is for MySQL JSON type.
-type JSON interface {
-	getTypeCode() byte
-	encode(buffer *bytes.Buffer)
-	// String returns JSON text.
-	String() string
-	// Type returns type string.
-	Type() string
+type JSON struct {
+	typeCode byte
+	i64      int64
+	str      string
+	object   map[string]JSON
+	array    []JSON
 }
 
 // CreateJSON creates a JSON from in. Return nil if in is not a valid JSON type.
 func CreateJSON(in interface{}) JSON {
```

Comment:
JSON won't bigger than Datum, and we have many functions return Datum directly.
so I prefer return it on stack. What do you think?